Delegated Opt In Review

joe-p opened the pull request.

joe-p committed:
add back files for review

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+## Abstract
+This ARC contains is implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). The goal is to provide users a way to delegate asset opt-ins while having a high degree of control compared to a standalone logic signature without an application. The expectation is that a single instance of this ARC will deployed on Algorand networks. 
```
```suggestion
This ARC contains is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). The goal is to provide users a way to delegate asset opt-ins while having a high degree of control compared to a standalone logic signature without an application. The expectation is that a single instance of this ARC will be deployed on Algorand networks. 
```

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+This ARC contains is implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). The goal is to provide users a way to delegate asset opt-ins while having a high degree of control compared to a standalone logic signature without an application. The expectation is that a single instance of this ARC will deployed on Algorand networks. 
+
+## Motivation
+[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control how long the signature is valid for. Without an application to control this, signing the logic signature program would be irriversible. There also needs to be a standardized way to store and ...
```
```suggestion
[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control how long the signature is valid for. Without an application to control this, signing the logic signature program would be irreversible. There also needs to be a standardized way to store and read signatures for a given account so dApps and other users can take advantaged of delegated opt-ins.
```

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control how long the signature is valid for. Without an application to control this, signing the logic signature program would be irriversible. There also needs to be a standardized way to store and ...
+
+## Specification
+This is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). There is additional functionality provided by the applicaiton in this ARC in the methods of the application.
```
```suggestion
This is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). There is additional functionality provided by the application in this ARC in the methods of the application.
```

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+
+### Signature Storage
+
+`setOpenOptInSignature(byte[64],address,txn)void` and `setAddressOptInSignature(byte[64],address,address,txn)void` are methods for adding signatures to box storage for open opt ins and address opt ins, respectively.
```
```suggestion
`setOpenOptInSignature(byte[64],address,txn)void` and `setAddressOptInSignature(byte[64],address,address,txn)void` are methods for adding signatures to box storage for open opt-ins and address opt-ins, respectively.
```

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+
+### End Times
+
+`setOpenOptInEndTime(uint64)void` and `setAddressOptInEndTime(uint64,address)void` are methods for setting the end time of a signature for open opt ins and ddress opt ins, respectively. An endtime signifies when the delegated logic signature will no longer work. The time corresponds to the epoch time (seconds) returned by `global LatestTimestamp`. End times can be updated at any time to any value ...
```
```suggestion
`setOpenOptInEndTime(uint64)void` and `setAddressOptInEndTime(uint64,address)void` are methods for setting the end time of a signature for open opt-ins and address opt-ins, respectively. An endtime signifies when the delegated logic signature will no longer work. The time corresponds to the epoch time (seconds) returned by `global LatestTimestamp`. End times can be updated at any time to any value.
```

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+
+`setOpenOptInEndTime(uint64)void` and `setAddressOptInEndTime(uint64,address)void` are methods for setting the end time of a signature for open opt ins and ddress opt ins, respectively. An endtime signifies when the delegated logic signature will no longer work. The time corresponds to the epoch time (seconds) returned by `global LatestTimestamp`. End times can be updated at any time to any value ...
+
+### Opt Ins
```
```suggestion
### Opt-Ins
```

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+
+### Opt Ins
+
+`openOptIn(pay,axfer)void` and `addressOptIn(pay,axfer)void` are implementations of the [ARCX](https://github.com/algorandfoundation/ARCs/pull/229) interfaces. They both verify the MBR payment is sent to the account opting in and that it covers the ASA minimum balance requirement, which is stored in global storage. It also verifies the value of `global LatestTimestamp` is less than the set end ti ...
```
By aware, though I don't think it matters, that `global LatestTimestamp` is for the _last_ block.

So if the last block was at 9998, and the end time is 10000, but the next block gets added at 10002, then you can have an opt-in the occurs after the set endtime.

joe-p replied:
Yeah that's why I specified the usage of `global LatestTimestamp`. The two options are either `Round` or `LatestTimestamp`. I feel like the timestamp approach allows for a better UX even though there may be some slight deviations in the true end time/block.

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+## Rationale
+Box storage is used to store signatures indefinitely and [the verifier lsig](./contracts/verifier_lsig.teal) ensures the signature is always correct.
+
+End time functionality is provided to allow users to revert the effects of signign the logic signature programs without having to rekey. It also let's users open their account for opt-in delegations for a short amount of time without having to remember to manually undo it.
```
```suggestion
End time functionality is provided to allow users to revert the effects of signing the logic signature programs without having to rekey. It also lets users open their account for opt-in delegations for a short amount of time without having to remember to manually undo it.
```

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+
+The test cases test proper functionality of all methods, but there has been no extended effort in attempt to break the contract. Most of the functionality in the app and logic signatures is relatively simple, so the chances of unexpected bugs is relatively low.
+
+It should be made clear that signatures are stored mapped to the signer (`auth-addr`) and end times are mapped to the public Algorand address. This means when a signature is made known, every account with the signing account as the `auth-addr` can now be opted in to asset via the delegated logic signature. Every account must set the end times to their desire, because without an end time set the s ...
```
```suggestion
It should be made clear that signatures are stored mapped to the signer (`auth-addr`) and end times are mapped to the public Algorand address. This means when a signature is made known, every account with the signing account as the `auth-addr` can now be opted-in to assets via the delegated logic signature. Every account must set the end times to their desire, because without an end time set the signature will work regardless of the latest timestamp.
```

jannotti commented on contracts/artifacts/DelegatedOptIn.clear.teal:
```diff
@@ -0,0 +1,3 @@
+#pragma version 8
+int 1
+return
```
```suggestion
```

jannotti commented on contracts/verifier_lsig.teal:
```diff
@@ -0,0 +1,47 @@
+int 1
++
+gtxns NumAppArgs
+switch _ _ _ open_optin_data addr_optin_data
```
I'd recommend giving the `_` label a meaningful name.

joe-p replied:
Renamed to `invalid_arg_count` and move respective `err` branch up in [80cd20e](https://github.com/joe-p/delegated_optin_arc/pull/1/commits/80cd20eaafe9c4f0a649b8b80e511e45e37a9ce2)

jannotti reviewed (commented):
Just a quick pass. I need to get further up to speed before any substantive comments.

joe-p committed:
Fix typos from review

Co-authored-by: John Jannotti <jannotti@gmail.com>

joe-p committed:
update addr offset

joe-p committed:
_ -> invalid_arg_count

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+## Abstract
+This ARC contains is an implementation of [ARCX](https://github.com/algorandfoundation/ARCs/pull/229). The goal is to provide users a way to delegate asset opt-ins while having a high degree of control compared to a standalone logic signature without an application. The expectation is that a single instance of this ARC will be deployed on Algorand networks. 
```
I can see why this is a nice, useful app to use in combination with the ARC-50, but I don't see why it should be considered "special". The point of ARC-50 was to give users control - let them have an app that they trust to check opt-ins for them.  This is one of them, but if it's the only one, then the flexibility that ARC-50 promises is reduced to the flexibility that this particular implementation provides.

joe-p replied:
Yeah the idea isn't that this is the only ARC50 implementation, but it's a standardized one that every wallet/dApp in the ecosystem can integrate. Other people can make whatever implementations they want, but the more implementations there are the more fragmented the ecosystem can get.

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+
+### End Times
+
+`setOpenOptInEndTime(uint64)void` and `setAddressOptInEndTime(uint64,address)void` are methods for setting the end time of a signature for open opt-ins and address opt-ins, respectively. An endtime signifies when the delegated logic signature will no longer work. The time corresponds to the epoch time (seconds) returned by `global LatestTimestamp`. End times can be updated at any time to any valu ...
```
Does setting the endtime to 0 delete the box?  Does the MBR get recovered somehow?

Actually, I'm assuming that the non-existence of an endtime means there is no endtime? So perhaps setting the endtime to MaxUint64 should delete the box and recover the MBR.

I suppose that really I'm asking for a way to delete all of the storage and have it mean "I want out of this whole business, give me my MBR back".

joe-p replied:
Right now there is no MBR recovery mechanism. I figured it added extra complexity for such a small amount of ALGO it wasn't necessarily worth it. Open to the idea of MBR recovery though.

jannotti replied:
I'm doubling down on my dislike of storing logicsigs on chain without a deletion mechanism.  If I say I want to allow someone to opt me in, I should just be able to remove that.

I suppose I'm really confused by the specific sender mechanism. What's the use case that makes it worth enshrining this idea?  Why not just "allow all, deny all"? (and by that I mean let's also scrap endtime.  just allow revocation.)

jannotti replied:
Ok, now I see. You are just storing the signature.  That seems very reasonable.

But it also means that using one box for the signature and the endtime is easy. They are both fixed size.

joe-p replied:
Assuming we still want to maintain the current design goals, we actually couldn't make it one box. The signatures are stored per the resepective public key, but the end times are stored per address. This means one can have multiple accounts rekeyed to a single auth account and have different rules set for each one. 

For example, I could have a single hardware wallet that is the auth addr for a "DeFi Account" and an "NFT Account". The "DeFi Account" delegates opt-ins to DeFi apps and "NFT Account" delegates opt-ins to NFT marketplaces, but not the other way around. This can currently be done because I can set the end times to be unique for each account.

joe-p replied:
I should mention the downside of this approach is it makes fresh rekeyed accounts open to any opt-ins before setting the end times, but in general I think this is reasonable tradeoff for having per-account control. Open to other opinions though.

jannotti replied:
> The signatures are stored per the resepective public key, but the end times are stored per address. This means one can have multiple accounts rekeyed to a single auth account and have different rules set for each one.
> 
> For example, I could have a single hardware wallet that is the auth addr for a "DeFi Account" and an "NFT Account". The "DeFi Account" delegates opt-ins to DeFi apps and "NFT Account" delegates opt-ins to NFT marketplaces, but not the other way around. This can currently be done because I can set the end times to be unique for each account.

I think it's probably a mistake to have this separation. Today, rekeying two accounts to the same key means that they are under the exact same authority.  It seems like a bad idea to introduce a mechanism that changes that model. If this becomes popular, we then have two ways of thinking about rekeying, and users must understand both.

joe-p replied:
Yeah true that's a really good point. Removing this means we can just use the existence of the signature boxes to signify whether delegated opt ins are permitted or not. We can also remove the time component and if people really want an automated disable, that can be done at a higher level.

jannotti commented on README.md:
```diff
@@ -0,0 +1,84 @@
+| Global | "assetMBR" | Stores the ASA MBR |
+| Box | "s-" + `auth-addr` | Mapping of signer to open opt-in signature |
+| Box |  "e-" + `auth-addr` | Mapping of signer to open opt-in end time |
+| Box | "s-" + `sha256(signer,sender)` | Mapping of the hash of the signer address plus sender address to address opt-in signature |
```
Using the hash, rather than concatenation saves little space (and MBR), but I'd like to understand the intended use, because it also makes some things harder. If an intended use is "allow an exchange to opt me in as needed", using the hash makes it impossible for the exchange to troll through the boxes to learn who it can opt-in for.  (It can check specific addresses, but not find out all the addresses is can opt-in.). To help with that, you might emit an event when such a box is set, which, one presumes, will be indexed by interested parties.

joe-p replied:
The hash is used because there are two key-value pairs that use the `(sender,signer)` tuple as a key (end times and signature), thus a prefix is needed to differentiate them which results in more than 64 bytes. 

There could be a single box that stores both the signature and the end time under a single key, but that does add some extra complexity to decoding that data structure (vs just checking box existence).

jannotti replied:
If the signature box got to be the one named by the concatenation, then you'd enable the kind of lookup I'm describing.

joe-p replied:
Ah true didn't think of that. Will update so the key for signatures is just the non-hashed `(sender,signer)` tuple

jannotti commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,530 @@
+	int 1
+	return
+
+setSigVerificationAddress:
+	proto 1 0
+
+	// ./contracts/delegated_optin_app.algo.ts:53
+	// assert(this.txn.sender === this.app.creator)
+	txn Sender
+	txna Applications 0
+	app_params_get AppCreator
+	assert
+	==
+	assert
+
+	// ./contracts/delegated_optin_app.algo.ts:54
+	// assert(!this.sigVerificationAddress.exists())
+	txna Applications 0
+	byte "sigVerificationAddress"
+	app_global_get_ex
+	swap
+	pop
+	!
+	assert
+
+	// ./contracts/delegated_optin_app.algo.ts:55
+	// this.sigVerificationAddress.set(lsig)
+	byte "sigVerificationAddress"
+	frame_dig -1 // lsig: address
+	app_global_put
+	retsub
```
Could the verify address be hardcoded at creation time and avoid a lot of audit surface?

(I separately want to have a discussion about why we're verifying them at all, but I'll a better place to start that conversation.)

joe-p replied:
Unfortunately not because the verify program requires the programs of the opt-in lsigs, which require the app ID of this application. Although I suppose the deployment process could be to deploy an `int 1` approval program and then update the program after the fact with the hardcoded address.

jannotti replied:
Oh, because it doesn't insert the app id at runtime?  You could make it do that. It'll be invoked in combo with this app, so it could grab the app id, insert it into a template, and verify.

joe-p replied:
What "runtime" are you referring to? The lsig needs to have the app ID hardcoded otherwise the user will let ANY app approve opt-ins, which is not desirable.

jannotti replied:
Right now, `verifier_lsig` is confirming that it's looking at a signature against `TMPL_OPEN_OPTIN_DATA` which I guess is the lsig bytecode, right? You could replace some bytes in `TMPL_OPEN_OPTIN_DATA` with the app id before calling `ed25519verify_bare`

Actually, I'm totally confused, it seems.  What the heck _is_ `TMPL_OPEN_OPTIN_DATA` and `TMPL_ADDRESS_OPTIN_DATA`. I thought it was bytecode, but I no longer think it is.

It's possible I have seriously misunderstood how this thing works.

joe-p replied:
No you're right. It's the data that's signed when a user signs an lsig, which is the lsig tag ("Program") + lsig bytecode. 
If the verifier program needs to know the app ID, then we're back to square one where the `verifier_lsig` program depends on the app ID thus the `verifier_lsig` needs to be created after the app.

jannotti commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,530 @@
+	int 1
+	return
+
+updateAssetMBR:
```
This is a clever trick to future-proof for MBR change on opt-in.

I'd rather give you `global AssetOptInMinBalance`. Would that eliminate this?

joe-p replied:
Yes that would be very useful. Exposing as much of the parameter tables as possible in general would be great.

jannotti replied:
I suppose I've avoided it in the hopes that programs would always use the pattern "Get MBR, do something, compute delta". But here you're trying to determine what a transaction would do even though that transaction already happened, at the top-level. You don't have a chance to get the mbr before and after. (So that's what you do, artificially, in this method.)

joe-p replied:
Yeah I think if there were to be an opcode like that it would only be useful in a scenario like this. Not sure if it'd be worth waiting for the next AVM version just for this functionality.

jannotti commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,530 @@
+	proto 3 0
+
+	// ./contracts/delegated_optin_app.algo.ts:99
+	// assert(verifier.sender === this.sigVerificationAddress.get())
+	frame_dig -3 // verifier: txn
+	gtxns Sender
+	byte "sigVerificationAddress"
+	app_global_get
+	==
+	assert
```
This seems to be as good a place as any to ask this: Why go through the trouble of verifying these lsigs?  It's a lot of complexity, and as far as I can tell, doesn't change any basic assumptions.  If the caller wants to make their sig unusable, so what?  The worst they can do is cause a transaction to fail, which costs callers nothing. 

I haven't looked at that code yet, but it presumably does a pretty opaque thing where it reconstructs bytecode by concatenation and does a big hash on it?  That's a pretty big audit surface to justify.

joe-p replied:
The current implementation allows *anyone* to upload the signature for a given address provided the signature is valid. If the app doesn't verify the lsigs, then it would have to implement logic that only allows the signer to put their signature in box storage. This would be counter-productive to one of the goals of reducing the onboarding friction of 0-ALGO accounts with an ASA.

jannotti replied:
Oh, wait, maybe I should have looked before I spoke. Is it verifying the code, or just the signature on it?

In the 0-algo case, who is generating the signed lsig?

jannotti replied:
Now that I see you're only putting the signature itself on-chain, ignore most of this.

But walk me through the 0-algo case. I don't get that. If someone is creating accounts, and therefore has keys, they could fund the account and perform the `set` transaction. Why are they willing and able to fund this thing, but not the account?

jannotti replied:
In fact, I *think* they don't even have to fund the account. If they have keys, they could make the transaction to `set` the sig, by calling this app, without even funding the account, just creating the app call transaction and putting it into a group with their own account which pays the fees.

joe-p replied:
Its just a matter of reducing the friction for getting the signature stored. The usefulness of this ARC is fairly dependent on the discoverability of signatures, so making it less clicks for the user to make their signature discoverable seems like it's worth the extra complexity of verification. In most scenarios (user interacting with an application), it's much easier for the app to programmatically set the sig box rather than having the user manually do it.

jasonpaulos replied:
> In most scenarios (user interacting with an application), it's much easier for the app to programmatically set the sig box rather than having the user manually do it.

If the user has to manually sign the lsig in the first place, I don't see why also having them sign and send a txn makes much of a difference.

And about the 0 algo case, it seems like users (those who want to opt-in others accounts) need to be aware of whether that account has 0 algos or not, since the MBR increase will be higher if they have to satisfy the base account MBR in addition to the asset MBR. This seems very annoying.

jannotti replied:
If the app can do that, it must have the user's keys, right?  If so, it can set the signature by sending a transaction on the user's behalf. If it doesn't have the keys, how did it generate the signature it is supplying to the call?

It's hard for me to see how the signature can be generated (by the app), but a transaction can't be sent (by the app).

joe-p replied:
The signature is still being created by the user. The difference is that the user uploading their own signatures means more clicks for them. They have to:

1. Create their account 
2. Sign the lsig
3. Share their address and signature
4. Wait for the app to respond with an atomic transaction that uses fee pooling from an account the app custodies
5. Sign the transaction

This is an obviously worse experience for the user because they have to sign the transaction and it's more complex for the app because they have to form the atomic transaction rather than simply calling the app with the signature.

jannotti replied:
It's not obvious to me that's worse because I don't know which of these operations are supported in a natural way with any existing wallets.  Do any wallets support signing lsigs? or arbitrary hunks of data, supplied by a dapp (<- I hope not!)?  If not, doesn't that kill both approaches?

joe-p replied:
No they don't currently, but [ARC47](https://github.com/algorandfoundation/ARCs/pull/226) will enable wallets to sign known lsigs. A wallet *could* also abstract away the signing of the open opt-in delegation lsig and just have it be a radio button during account creation.

jannotti commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,530 @@
+	>=
+	assert
+
+	// if0_condition
+	// ./contracts/delegated_optin_app.algo.ts:117
+	// this.openOptInEndTimes.exists(optIn.assetReceiver)
+	byte "e-"
+	frame_dig -2 // optIn: axfer
+	gtxns AssetReceiver
+	concat
+	box_len
+	swap
+	pop
+	bz if0_end
```
Rather than check for existence with `box_len`, and then use `box_get` but through away the existence bit, why not just use `box_get` and use that bit?

joe-p replied:
The TEAL is generated by TEALScript, which has seperate `get` and `exist` functions and no direct exposure of `box_get` without popping one of the values. Compiler optimization can eventually be made to account for this sort of pattern but in the meantime the TEAL would need to be manually edited to optimize this.

jannotti replied:
This isn't the only place where this extra complexity is coming up.  If this is going to be audited, they are going to have to audit the TEAL, as they are not going to be able to trust all of TEALScipt compilation without auditing _it_.

There are a lot of places where we're creating a lot more audit work by auditing compiler output. (things like confirming that each methods checks `OnCompletion` instead of doing it in one place, the extra `callsubs` and "trampolines" for each method, etc.)

joe-p replied:
I question how much extra "audit work" an extra `box_len` or `callsub` really adds. I'm not super familiar with the auditing process, but I know we got a reasonable quote and timeline for ARC12 which was TEALScript-generated and more complex.

Regarding OnCompletion, it's been on the TODO list to have only one entry point per OnCompletion in the generated TEAL from TEALScript, so this is something that can be updated.

jannotti reviewed (commented):
I don't really like the idea of using boxes for "storage" in the sense of "this is something to be grabbed and used by an off-chain program" as is done with logicsigs here.  Just storing the hash of the logicsig should be sufficient in combination with off-chain storage.

Alternatively, I suppose you could store *just* the signature in the boxes, since the bytes are deterministic (right? I think so) and wallets could reconstruct them. I suppose that's placing a bit more burden on them.

jasonpaulos commented on contracts/address_optin_lsig.teal:
```diff
@@ -0,0 +1,52 @@
+txn AssetReceiver
+txn Sender
+==
+assert
```
This file is missing some checks:
1. Checking that the rekey to field is empty
2. Checking that this is an axfer txn. I suppose this isn't strictly necessary because this code implies it being true (`txn AssetReceiver; txn Sender; ==; assert`), but I'd feel more comfortable with an explicit check.

Same applies to `open_optin_lsig.teal`

joe-p replied:
Fixed in [76ca460](https://github.com/joe-p/delegated_optin_arc/pull/1/commits/76ca46084a97a3f393f79c200aa6a9a65738f43d)

jasonpaulos commented on contracts/delegated_optin_app.algo.ts:
```diff
@@ -0,0 +1,186 @@
+  setOpenOptInSignature(sig: byte64, signer: AuthAddr, verifier: Txn): void {
+    assert(verifier.sender === this.sigVerificationAddress.get());
+
+    this.openOptInSignatures.set(signer, sig);
```
Who pays for the box?

joe-p replied:
Can be anyone really. I don't see a reason why it'd need to be a specific account.

jasonpaulos commented on contracts/delegated_optin_app.algo.ts:
```diff
@@ -0,0 +1,186 @@
+    assert(optIn.assetReceiver === mbrPayment.receiver);
+    assert(mbrPayment.amount >= this.assetMBR.get());
+
+    // If endTimes box exists, openOptIn that the opt in is before the end time
+    if (this.openOptInEndTimes.exists(optIn.assetReceiver)) {
+      assert(this.openOptInEndTimes.get(optIn.assetReceiver) > globals.latestTimestamp);
+    }
```
What is the benefit of supporting end times vs just allowing the account to immediately stop supporting the operation? E.g. if this check could instead be `assert(this.openOptInSignatures.exists(optIn.assetReceiver))`.

Asking because:
* Our timestamps are not very accurate anyway
* It seems like boxes containing expired signatures/timestamps must continue to exist forever. If you instead make the signature box existence indicative of whether the account wants to receive opt-ins or not, they'll clean themselves up automatically.

joe-p replied:
The existance of the signature is not used to determine functionality because I wanted control over functionality to be per-address, rather than per-auth-addr. Rationale explained [here](https://github.com/joe-p/delegated_optin_arc/pull/1#discussion_r1284695447). 

Since a separate box is needed anyways for this, I figured might as well use timestamps to reduce clicks in the event the user knows they only want to delegate for a specific amount of time.

jasonpaulos replied:
What do you mean by per-address, rather than per-auth-addr? Are you talking about the `addressOptIn` path? If so this suggestion applies just as well to that case:

```typescript
assert(this.addressOptInSignatures.exists({ signer: optIn.assetReceiver, sender: this.txn.sender } as SignerAndSender))
```

joe-p replied:
Let's say I have three accounts: account A and account B which both are rekeyed to account C. If I sign the lsig with account C, then anyone can use that signature for account A, B or C. But... maybe I only want to delegate opt-ins for account C, and disable delegated opt-ins on account A and B.

jasonpaulos replied:
Oh I see, you're saying that the key for `openOptInSignatures` represents auth addresses, NOT accounts. So in your example, with accounts A and B rekeyed to C, A and B would both use the signature from `openOptInSignatures[C]`. Meaning we can't use only the existence of boxes from `openOptInSignatures` to indicate opt-in status, since it's a many-to-one mapping, and each account should be able to control that individually.

I have two alternatives suggestions then:
1. Make each `openOptInSignatures` control exactly one account, not an auth address that can back multiple accounts. You'd need to augment the values here, since it's necessary to store the actual signing key for the signature, since it can now be different from the account's key (e.g. vale type could be a tuple of `(bytes32, bytes64)`).
    * Pros:
      * Boxes get cleaned up when they're no longer necessary
    * Cons:
      * Slightly larger box keys to hold info which most of the time won't be necessary (since I assume most accounts aren't rekeyed)
3. Keep the existing convention of `openOptInSignatures`, but instead of having `openOptInEndTimes = new BoxMap<Address, uint64>` control opt-in status, use something like `openOptInEnabled = new BoxMap<Address, null>` (I don't know how you'd represent a zero-length box value in tealscript, but that's what I'm trying to say with `null`). `openOptInEnabled` would control whether an opt-in is allowed for an account: if and only a box exists for that account, it's allowed.
    * Pros:
      * `openOptInEnabled` boxes will get cleaned up when no longer needed.
    * Cons:
      * `openOptInSignatures` boxes won't get cleaned up when no longer needed

jasonpaulos commented on contracts/verifier_lsig.teal:
```diff
@@ -0,0 +1,47 @@
+txn RekeyTo
+global ZeroAddress
+==
+assert
```
Why doesn't the verifier make any claim about the transaction being a certain type?

Sure if the plan is for it to own 0 algos, that's not as important, since it has nothing to drain anyway, but this enables the following weird things to happen:
* Anyone can make this lsig account create assets/apps (weird but likely harmless?)
* Anyone can make this lsig account call into an app while interacting with the `DelegatedOptIn` contract. This is something to be weary of, as this app call could read intermediate state from that contract, or even invoke the contract _again_ from an inner txn group in an attempt to mess up its state or logic. Best not to give any bad actors this lever in my opinion.

joe-p replied:
In [76ca460](https://github.com/joe-p/delegated_optin_arc/pull/1/commits/76ca46084a97a3f393f79c200aa6a9a65738f43d) I asserted that it's a pay.

joe-p committed:
dont hash or prefix sig key for discoverability

joe-p committed:
more lsig checks, update tests

algoanne commented:
hey @joe-p, we (myself, @jasonpaulos and @jannotti ) had a live chat about this. Would you be open to scoping down to the simplest use case: someone who has a funded account wants to allow anybody to opt them in to assets. 

This is likely also the most common use case. We could get it out there with the wallets and it would be immediately useful. We could then add the potentially more complex features like address-based permissions in the future, perhaps along with any updates based on feedback from this first main feature. 

In other words, the following changes: 
- only the "open opt in".
- only the person signing can call the app to register their delegated logic sig. 
- no timeout / timestamp mechanics. Instead, the presence of the delegated logic sig in the app serves as the on/off. 

The resulting design should be fairly simple and therefore easier to feel confident about, while maintaining most of the usefulness. In our discussions, most of the complexity comes from the additional requirements. 

Another way of thinking about this MVP: enabling wallets to add a toggle for "let anybody send me assets".

joe-p commented:
The main rationales for having all of these features in this implementation are cost and complexity. For every contract we deploy, the foundation needs to pay for an audit. Paying for multiple audits is obviously going to be more expensive and we might not have the budget for that. This ultimately isn't my call so we'd need to have this conversation with Alessandro (and/or John Woods). 


The features in this contract alone don't add any complexity by coexisting. Separating them does nothing but ADD complexity. Yes this contract is more complex, but it's no more complex than the sum of its parts. 

If we had separate contracts deployed at separate times: 

- Applications would have to query box storage for multiple contracts when sending someone an asset.
- A user might have opt-ins enabled in one app but not another. That could be rather confusing. 
- Some apps will implement this, but maybe not future standards, which would add more friction and confusion in the ecosystem as a whole. 


> no timeout / timestamp mechanics. Instead, the presence of the delegated logic sig in the app serves as the on/off.

I'm open to changing the implementation but don't want to sacrifice address-specific opt-ins. I also would want some more rationale other than "it's more simple". Having per-address control seems like a valuable feature to me. We were also okay with proceeding with ARC12, which was certainly more complex than this. I suppose it's just not clear to me what makes something "too complex".

algoanne commented:
@joe-p I am driving from the typical wisdom of agile development. The skateboard -> scooter -> motorcycle -> car method, which proffers that you should start with the simplest useful thing (the "MVP") and iterate from there. This allows you both to get value out there as quickly as possible, and learn from version 1 to inform version 2, ultimately exposing yourself to less risk and leading to a better long-term outcome. 

In terms of time to value, it will be easier for the ecosystem to adopt the simple open delegated opt-in: it only requires work from wallets, no involvement from dapps. It's also easier to explain and understand for everybody, making it more likely to get buy-in. 

In terms of risk: there is a lot of risk in this work, risk essentially on the level of a protocol change - this code will be used by everybody in the ecosystem. If there is a bug in the code or a problem in the design (something that we're not accounting for that could surface later), it will be a huge pain for all of us. 

And just anecdotally for risk, I was listening to JJ and Jason discuss the code and going "wow this is complicated/subtle" to each other and it made me nervous. 

To be clear: I do think it will _probably_ be useful to have the address-based opt-in delegation. I am only offering my Product perspective on what I think is the best approach to get there. 

I would hope that an external audit of the simpler contract would be cheaper, but I know these things are not always logically priced. If not, then yes it depends on the opinion of the powers that be at the Foundation. In terms of our audit from Inc engineers, we are (more) happy to audit two separate approaches. 

I'm not going to die on this hill, and it's not my product. This is just my recommendation.

joe-p commented:
I'm not totally against it, but I think we should really think about the pros/cons of simplifying the initial deployment.

Pros

1. Simpler functionality means less chance for bugs
2. Maybe we find out people don't want per-address control
3. Potentially cheaper initial audit

Cons

1. Seperating functionality could result in fragmented adoption in ecosystem
2. Seperate implementations could complicate integration decisions for wallets and applications
3. Not having a address-specific opt-in could result in applications implementing their own lsig-based solution
4. Net amount spent on audits likely increased

I think con 3 is something in paticular that we have to think about. Once wallets support templated logic signatures, I imagine apps will start pushing for a opt-in lsig. If this lsig does not use an application like this implementation, then users will have no way to undo the lsig without rekeying. There are also multi-chain implications that users might not be aware of. For example, signing an lsig for opting in on Algorand mainnet will also work on Voi.

jannotti commented:
To force myself to understand my own claims, I removed all of the things we've talked about cutting:
1. No support for sender specific authority
2. No endtime, signature existence determines approval
3. No direct zero algo support - only sender can set their own signature
4. No Tealscript - removed a lot of callsub stuff (but retained all "high-level" comments and blank lines)

This eliminates the verification teal script (54 lines) and the main approval script goes from 522 lines to 145.  83 of that 145 is the `updateAssetMBR` machinery, which we should replace with `global AssetOptinMinBalance`. 

So the audit goes from two tealscripts of 576 lines, to one script of 61 lines.  I believe there would be just 45 lines of meat:
```
setSignature:
        txn Sender
        txn ApplicationArgs 1
        dup
        len
        bnz really_set
        box_del                 // delete if set empty sig
        return
really_set:
        box_put
        int 1
        return

optIn:
        txn GroupIndex
        int 1
        -
        store 0                 // optIn: axfer
        txn GroupIndex
        int 2
        -
        store 1                 //  mbrPayment: pay

        // assert(optIn.assetReceiver === mbrPayment.receiver)
        load 0 // optIn: axfer
        gtxns AssetReceiver
        load 1 // mbrPayment
        gtxns Receiver
        ==
        assert

        // assert(mbrPayment.amount >= this.assetMBR.get())
        load 1 // mbrPayment: pay
        gtxns Amount
        global AssetOptInMinBalance
        >=
        assert

        // signature exists
        load 0 // optIn: axfer
        gtxns AssetReceiver
        box_get
        assert
        len
```

jannotti commented:
Getting back to actually auditing for safety, I think we should aim for both of the following to be true, even though we can imagine "splitting" some of the checking responsibility. If we aim for assurance in both directions, we will feel very safe.

1. Looking only at the approval app, everyone should feel safe signing _any_ logic sig that requires that app in the next transaction _regardless_ of whatever other checks are in the logic sig being signed.

2. In the other direction, looking _only_ at the logic sig the user signs, such a user should be confident that it can't be used to do anything except opt them into an asset with the MBR paid to them, regardless of what the app says. (They also _like_ the property that the app blocks opt-ins unless the signature is still present in a box, but if all else fails, they want that logic sig to be no more dangerous than "Allow compensated opt-ins".)

As I read the app, it does not have these checks. It allows the axfer to be a clawback, for example.  Suppose a user, C opts into this facility and C is also the clawback address for some asset, A.  I am a bad guy, E, who wants to use C's power to steal some A from random guy R.

According to the _app_ I can send a payment to E (me), followed by a clawback of A from R to E, followed by this app.  The app only confirms that the AssetRecipient and the payment go to the same place. They do, ME!

I recognize that the lsig protects us. It checks the AssetAmount == 0, and Sender == AssetReceiver, but I'd like to see that check in the app (and AssetSender == ZeroAddress - no clawbacks).

Meanwhile, only the app checks that there's a pay that covers MBR. In part, that's because it uses the inner opt-in trick. But I'd like to see that check also appear in the lsig, and I'm very willing to add the necessary `global` to make that possible.

joe-p commented:
> No direct zero algo support - only sender can set their own signature

I'm still not convinced that this is worth removing... mainly for peer-to-peer transactions. In general, it increases the amount of clicks and potentials areas of failure for the initial asset on-boarding process. It also means the user being onboarded has to have an active internet connection and access to their keys when first receiving an asset. Making an account on your ledger and then later sharing a QR code is no longer viable if you don't carry the ledger with you. To do this programmatically, the user would have to sign yet another lsig, which would need to include fv/lv and a lease, which means wallets or applications would need to generate this at the time of the transaciton, which could eventually become stale. 

> Looking only at the approval app, everyone should feel safe signing any logic sig that requires that app in the next transaction regardless of whatever other checks are in the logic sig being signed.

I'm confused as to why this is necessary. Any meaningful transaction is going to include both the logic signature and the app, which is hard coded into the logic signature. Without the app you don't have a functioning logic signature and without the logic signature you are burning ALGO to clal the app which does nothing of value. It doesn't make any sense for anyone to use the app without the lsig.

jannotti commented:
> > No direct zero algo support - only sender can set their own signature
> 
> I'm still not convinced that this is worth removing... mainly for peer-to-peer transactions. In general, it increases the amount of clicks and potentials areas of failure for the initial asset on-boarding process. It also means the user being onboarded has to have an active internet connection and access to their keys when first receiving an asset. Making an account on your ledger and then later sharing a QR code is no longer viable if you don't carry the ledger with you. To do this programmatically, the user would have to sign yet another lsig, which would need to include fv/lv and a lease, which means wallets or applications would need to generate this at the time of the transaciton, which could eventually become stale.

Can you go through _exactly_ the _entire_ scenario you'd like to support?  I see these weird combinations like "I have ledger, that costs real money, but care a lot about security, so I don't carry it, but I have no algos, and I want a QR code..." I can't follow it.

To be clear, I _did_ eventually see why you were worried about an extra click previously. You felt a dapp would be sending the logicsig to the wallet for signing, and _then_ the dapp would construct a txngroup to perform the setSig. So you didn't want that txngroup to need to be signed as well. 

But here, I think this stripped down version is good for exactly one thing - your _wallet_ wants to offer a checkbox that says "I accept airdrops".  You cannot check this box until you have 0.1 + epsilon algos.  Once you have that, everything is done internally to the wallet - no dapp is presenting a logicsig to sign - there's just the single builtin logicsig that enables these opt-ins. Your wallet signs the logicsig and makes the txngroup for you all at once. (It would probably have to interact with a ledger twice, if that's how you are doing things, but it would do them as "together" as possible. I don't think we should be designing around the problem "ledgers are too hard to use to sign a transaction".). I have no idea what logicsig signing is going to be like with a ledger, however. Do they offer the ability to sign a blob? That problem exists for both designs.

jannotti commented:
> > Looking only at the approval app, everyone should feel safe signing any logic sig that requires that app in the next transaction regardless of whatever other checks are in the logic sig being signed.
> 
> I'm confused as to why this is necessary. Any meaningful transaction is going to include both the logic signature and the app, which is hard coded into the logic signature. Without the app you don't have a functioning logic signature and without the logic signature you are burning ALGO to clal the app which does nothing of value. It doesn't make any sense for anyone to use the app without the lsig.

As I said, it is not necessary. One could be willing to analyze all the moving parts at once. But I think it's pretty easy to make it easier to know we've got it right by make each part independently safe.  It's maybe 3 or 4 extra asserts on both sides. It makes it much easier to state the expectation from each part.  For example, looking at your original PR, I had to ask myself "Why is the delegated logicsig checking the sender, can't the app do that?"  I don't want to ask any such questions. I want to know that I'm safe if we've got the analysis right on either piece.

joe-p commented:
> Can you go through exactly the entire scenario you'd like to support? I see these weird combinations like "I have ledger, that costs real money, but care a lot about security, so I don't carry it, but I have no algos, and I want a QR code..." I can't follow it.

Someone has a ledger and actively uses other chains. They hear there's an ALGO event nearby, so they create a wallet, leave the ledger at home, and go to the event with just their phone. Event in this case could be a crypto conference or something like a concert that uses NFT ticketing at the door.  

For some context, the way we currently onboard new users at events is print physical paper cards with a QR code containing mnemonics that is already seeded with some ALGO and opted into whatever relevant assets. 

> Once you have that, everything is done internally to the wallet - no dapp is presenting a logicsig to sign 

This is true for both designs though. The wallet just needs to generate the sig once upon account creation or from a click in the settings menu and then it's simply a matter of how it's communicated to other parties. Wallets can push the signature to something like DynamoDB, and then anyone can read that signature and use it. There is really no reason the signature *needs* to be stored on-chain at all. That data is never used on-chain. Box storage is simply the cheapest, low-maintenance and decentralized way to store signatures long-term.

> I have no idea what logicsig signing is going to be like with a ledger, however. Do they offer the ability to sign a blob? That problem exists for both designs.

Not sure if logic signature signing is supported, but I know arbitrary message signing is definitely not supported.

jannotti commented:
> > Can you go through exactly the entire scenario you'd like to support? I see these weird combinations like "I have ledger, that costs real money, but care a lot about security, so I don't carry it, but I have no algos, and I want a QR code..." I can't follow it.
> 
> Someone has a ledger and actively uses other chains. They hear there's an ALGO event nearby, so they create a wallet, leave the ledger at home, and go to the event with just their phone. Event in this case could be a crypto conference or something like a concert that uses NFT ticketing at the door.

So they are out of luck in both designs, right?  They need to get the logicsig signed by this new account. But they are not experienced enough to have done that at home with their ledger.  If they don't have their keys with them (no ledger) then they can't sign the logicsig OR send the txn to chain once they arrive at event.

> For some context, the way we currently onboard new users at events is print physical paper cards with a QR code containing mnemonics that is already seeded with some ALGO and opted into whatever relevant assets.

If we're willing to do that, then they can send the algos from that card to their pre-existing account. But I still don't understand how they can do anything without pre-signing that logicsig, which they surely won't be experienced enough to do ahead of time.

> > Once you have that, everything is done internally to the wallet - no dapp is presenting a logicsig to sign
> 
> This is true for both designs though. The wallet just needs to generate the sig once upon account creation or from a click in the settings menu and then it's simply a matter of how it's communicated to other parties. Wallets can push the signature to something like DynamoDB, and then anyone can read that signature and use it. There is really no reason the signature _needs_ to be stored on-chain at all. That data is never used on-chain. Box storage is simply the cheapest, low-maintenance and decentralized way to store signatures long-term.

So for this to work you're imagining Pera will build in some special bootstrapping mode where this user *will* sign the logicsig at home with their ledger, then bring the signed bytes in on a thumb drive or printed QR or something?  It all seems way easier to make a temporary wallet at the event using Pera, get some algos at the event, and if they want, send the algos and the assets to their secure wallet when they get home.  Ledger seems like a red herring.

> > I have no idea what logicsig signing is going to be like with a ledger, however. Do they offer the ability to sign a blob? That problem exists for both designs.
> 
> Not sure if logic signature signing is supported, but I know arbitrary message signing is definitely not supported.

So this is all a bit hypothetical with respect to ledgers, isn't it?

joe-p committed:
initi review updates

joe-p committed:
update TEALScript

joe-p committed:
doesn't allow opt-ins test

joe-p committed:
rm authAddr resolution

joe-p committed:
more validation in lsig

joe-p committed:
use >= for mbr payment amount

joe-p committed:
rm asset MBR from app

joe-p commented:
I just updated the implementation based on what we discussed.

- Removed address-specific opt-in delegation
- Removed asset MBR calculation/verification from app
- Txn validation in logic sig
- Remove end time and just use signature existence

Key missing part is `global` field for asset opt-in MBR. 

Also, the application currently does no verification of the other txns in the group. With the additional of the aforementioned global field, the app itself won't need to do any verification thus I feel like it makes more sense to just keep everything in the logic sig, but open to other opinions.

jasonpaulos commented on README.md:
```diff
@@ -2,7 +2,7 @@
 
 ## Motivation
-[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control how long the signature is valid for. Without an application to control this, signing the logic signature program would be irreversible. There also needs to be a standardized way to store and ...
+[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control whether the lsig is still functionality or not. Without an application to control this, signing the logic signature program would be irreversible without rekeying. There also needs to be a s ...
```
```suggestion
[ARCX](https://github.com/algorandfoundation/ARCs/pull/229) provides a standard for delegated asset opt-ins, but there are some UX problems that need to be addressed. First, there needs to be a way to control whether the lsig is still functional or not. Without an application to control this, signing the logic signature program would be irreversible without rekeying. There also needs to be a standardized way to store and read signatures for a given account so dApps and other users can take advantaged of delegated opt-ins.
```

small typo

jasonpaulos commented on README.md:
```diff
@@ -0,0 +1,65 @@
+## Rationale
+Box storage is used to store signatures indefinitely and [the verifier lsig](./contracts/verifier_lsig.teal) ensures the signature is always correct.
+
+End time functionality is provided to allow users to revert the effects of signing the logic signature programs without having to rekey. It also lets users open their account for opt-in delegations for a short amount of time without having to remember to manually undo it.
```
This should be updated

joe-p replied:
Updated in b69d790b048c5682fd8549e063b609651f01bb4c

jasonpaulos commented on README.md:
```diff
@@ -0,0 +1,65 @@
+
+The test cases test proper functionality of all methods, but there has been no extended effort in attempt to break the contract. Most of the functionality in the app and logic signatures is relatively simple, so the chances of unexpected bugs is relatively low.
+
+It should be made clear that signatures are stored mapped to the signer (`auth-addr`) and end times are mapped to the public Algorand address. This means when a signature is made known, every account with the signing account as the `auth-addr` can now be opted-in to assets via the delegated logic signature. Every account must set the end times to their desire, because without an end time set the  ...
```
same here

joe-p replied:
Updated in b69d790b048c5682fd8549e063b609651f01bb4c

jasonpaulos commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,165 @@
+*
+txn OnCompletion
++
+switch create_NoOp NOT_IMPLEMENTED NOT_IMPLEMENTED NOT_IMPLEMENTED NOT_IMPLEMENTED NOT_IMPLEMENTED call_NoOp
```
This is a bit weird. Maybe the `match` opcode would help?

Probably the simplest thing to do is just assert that the OnCompletion type is NoOp instead though.

edit: I guess this is a tealscript artifact? Perhaps it could also include a brief comment about what it's doing (creating slots for the product of cases).

joe-p replied:
Yeah this is how TEAL handles routing. I eventually want to make this an ARC so explorers can just parse the first couple lines of TEAL and if its `txn ApplicationID; int 0; >; int 6; *; txn OnCompletion; +; switch ...` follow the branch for delete/update to see if it's `err`, which would allow for explorers to show a binary flag on whether a contract is mutable or not.

joe-p replied:
Added a comment explaining what the code is doing and why in c5d72037dcc7c69de47a1e8a018707090a949b9e

jasonpaulos commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,165 @@
+	// preMBR = this.app.address.minBalance
+	txna Applications 0
+	app_params_get AppAddress
+	assert
```
You could use `global CurrentApplicationAddress` instead

joe-p replied:
Fixed in c5c2d9547c3da8bae8f1979f638c3d5192b5ae72

jasonpaulos commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,165 @@
+	err
+
+abi_route_setSignature:
+	byte 0x
```
Why is this empty string an argument? Is this how tealscript implements subroutine local variables?

joe-p replied:
Yeah TEALScript uses the frame for all local variables and arguments, so it needs to reserve the space in the frame for them, thus its pushes empty bytes for however many variables there are.

joe-p replied:
Added comments for clarity in c5d72037dcc7c69de47a1e8a018707090a949b9e

jasonpaulos commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,165 @@
+	swap
+	pop
+	assert
+	retsub
```
This function never checks the pay?

jannotti replied:
I believe he's counting on the signed logicsig to check.

I think they both should. It's just easier to understand if a reader of each program can reason without seeing the other.  Of course, you can't reason about *all* the security properties without seeing them in combination. 

In particular, reading the logicsig alone can't guarantee that the app handles revocations properly, but it can protect itself from anything except compensated opt-ins (spam).  Meanwhile, the app could be written in a way that if I know my logicsig requires the app's existence, I'm safe, regardless of what else the logicsig I'm signing does. I could imagine, for example, that if this was at the top:
```
// Verify called atomically with master app
AppCall
gtxns ApplicationID
int TMPL_DELEGATED_OPTIN_APP_ID
==
assert

// Verify the correct method is being called
AppCall
gtxnsa ApplicationArgs 0
method "delegatedOptIn(pay,axfer)void"
==
assert
```
Then there could be wallets that say "You're safe, you're being asked to signed a logicsig for an app/method that is on my allowlist of known trustworthy apps."

I know this is not strictly necessary, but we should be making auditing trivial and mistakes as impossible as we can.

I even think the app should check things the logicsig checks, like Fee=0, no re-key. Both should be as clearly harmless as can be.

joe-p replied:
Yeah I know this was originally discussed. I thought the reasoning for having checks in the app was because some checks needed to be in the app and some needed to be in the lsig, but now all checks can be in the lsig so it's less confusing.

I have no problem with adding checks to the app though so will add in the next update

joe-p replied:
I have moved app verification to the top of the TEAL in 75ceab28e1f95ec754bfe0b1452e27ace2806206 and added transaction verification to the app in 4701fd7d4b0204e167a92182d36262b8d48bcb4a.

jasonpaulos commented on contracts/artifacts/DelegatedOptIn.abi.json:
```diff
@@ -0,0 +1,55 @@
+    {
+      "name": "unsetSignature",
+      "args": [],
+      "desc": "Delete the signature from box storage.This will disable delegated opt-ins and return the box MBR balance",
```
This should definitely note that a surplus group fee is required, otherwise the pay txn with the MBR balance will fail.

joe-p replied:
Added in 7a559c552bca9c73b5c43bb303b14bf545d787e7

jasonpaulos commented on contracts/open_optin_lsig.teal:
```diff
@@ -0,0 +1,93 @@
+assert
+
+// Verify assetCloseTo is not set
+txn AssetCloseTo
```
While we're at it I'd also verify AssetSender is not set

jannotti commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,165 @@
+	int 1
+	return
+
+setSignature:
```
Maybe TEALScipt should emit the function signature here as a comment? (or next to the `proto` line).  Later `frame_bury` shows the thing being accessed, it would be nice to "declare" those names here.

joe-p replied:
Yeah makes sense to me. In general sounds like more comments in the emitted TEAL would be good.

joe-p replied:
Added header comments containing the signature along with the method/arg/return desc in c5d72037dcc7c69de47a1e8a018707090a949b9e

jannotti commented on contracts/artifacts/DelegatedOptIn.approval.teal:
```diff
@@ -0,0 +1,165 @@
+	retsub
+
+abi_route_delegatedOptIn:
+	// no dupn needed
```
What does this comment mean? And why wasn't it emitted at the start of the other `abi_` method, which also did not begin with `dupn`?

joe-p replied:
TEALScript uses the frame for local variables, so it needs to push values to the stack when there are local variables prior to calling the subroutine. `// no dupn needed` is basically saying that there doesn't need to any extra values pushed to the stack since there are no local variables. 

This comment will be improved along with more comments for clarity in a near-future update

joe-p replied:
Resolved in c5d72037dcc7c69de47a1e8a018707090a949b9e

jannotti commented on README.md:
```diff
@@ -0,0 +1,65 @@
+
+### Signature Storage
+
+`setSignature(byte[64],address,txn)void` is a method that allows a user to upload their signature to box storage
```
Is it really `setSignature(byte[64],address,pay)void`

(txn -> pay)

joe-p replied:
Updated in b69d790b048c5682fd8549e063b609651f01bb4c

jannotti commented on README.md:
```diff
@@ -0,0 +1,65 @@
+      ✓ works with valid signature and lsig (790 ms)
+    delegatedOptIn
+      ✓ works with valid lsig and method call (1453 ms)
+    unsetSignature
```
How about "revokeSignature" instead?

joe-p replied:
Yeah I like that better. Changed in 89e1515a7c69e88f5370735d8d516379549fbc77

joe-p committed:
Update README.md

Co-authored-by: Jason Paulos <jasonpaulos@users.noreply.github.com>

joe-p committed:
use "global CurrentApplicationAddress"

joe-p committed:
update README

joe-p committed:
unsetSignature -> revokeSignature

joe-p committed:
update revokeSignature desc

joe-p committed:
update TEALScript

joe-p committed:
verify txns in app call

joe-p committed:
move app verificaiton to the top

joe-p committed:
fix typo

joe-p committed:
update README

joe-p committed:
use verifyTxn for boxMBRPayment
